File: cloud-functions/python/code.py
import json

def deep_equal(a, b):
    if a == b:
        return True
    if type(a) != type(b):
        return False
    if isinstance(a, dict):
        if set(a.keys()) != set(b.keys()):
            return False
        for key in a.keys():
            if not deep_equal(a[key], b[key]):
                return False
        return True
    if isinstance(a, list):
        if len(a) != len(b):
            return False
        for i in range(len(a)):
            if not deep_equal(a[i], b[i]):
                return False
        return True
    return False

def python_code(request):
    # Handling CORS headers for preflight requests
    if request.method == "OPTIONS":
        headers = {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "GET, PUT, PATCH, POST, DELETE",
            "Access-Control-Allow-Headers": request.headers.get("access-control-request-headers"),
        }
        return ("", 204, headers)

    try:
        request_data = request.get_json()
        code = request_data["code"]
        question = request_data["question"]

        results = []

        for input_data in question["inputsOutputs"]:
            sandbox_locals = {}
            sandbox_locals["printed_output"] = []
            
            code_string = (
                "def execute_user_code():\n"
                f"  {code}\n"
                f"  globals()['res'] = solution({json.dumps(input_data['inputs'])})\n"
                "execute_user_code()"
            )

            def custom_print(*args):
                formatted_args = []
                for arg in args:
                    if isinstance(arg, str):
                        formatted_args.append(arg)
                    else:
                        formatted_args.append(json.dumps(arg))
                output = " ".join(formatted_args)
                sandbox_locals["printed_output"].append(output)
            
            sandbox_locals['print'] = custom_print

            exec(code_string, sandbox_locals)

            result = sandbox_locals["res"]
            printed_output = sandbox_locals["printed_output"]
            results.append({"result": result, "log": printed_output})

        assessment_result = {
            "inputs": [input_data["inputs"] for input_data in question["inputsOutputs"]],
            "outputs": [result["result"] for result in results],
            "expectedOutputs": [input_data["output"] for input_data in question["inputsOutputs"]],
            "isCorrect": [deep_equal(result["result"], input_data["output"]) for result, input_data in zip(results, question["inputsOutputs"])],
            "logs": [result["log"] for result in results]
        }

        headers = {
            "Access-Control-Allow-Origin": "*",
            "Content-Type": "application/json"}
        return (json.dumps(assessment_result), 200, headers)

    except Exception as e:
        import traceback
        traceback.print_exc()
        headers = {
            "Access-Control-Allow-Origin": "*",
            "Content-Type": "application/json"}
        return (json.dumps({"error": str(e)}), 200, headers)

File: cloud-functions/js/code.js
const vm = require("node:vm");
const process = require("process");
const { performance } = require("perf_hooks");

exports.jsCode = (req, res) => {
  res.header("Access-Control-Allow-Origin", "*");
  res.header("Access-Control-Allow-Methods", "GET, PUT, PATCH, POST, DELETE");
  res.header("Access-Control-Allow-Headers", req.header("access-control-request-headers"));
  if (req.method === "OPTIONS") {
    res.status(204).send("");
  } else {
    try {
      const { code, question } = req.body;

      console.log("req.body", req.body);

      const results = [];
      for (const input of question.inputsOutputs) {
        const sandbox = {
          results: null,
          executionTime: 0,
          memoryUsage: 0,
          process,
          performance,
          log: [],
          console: {
            log: (...args) => {
              let formattedArgs = args.map((arg) => (typeof arg === "string" ? arg : JSON.stringify(arg)));
              sandbox.log.push(formattedArgs.join(" "));
            },
          },
        };
        vm.createContext(sandbox);
        const script = `const startTimestamp = performance.now();
          const startMemoryUsage = process.memoryUsage().heapUsed;
          results = (function() {
          ${code}
          return solution(...${JSON.stringify(input.inputs)});
          })();
          const endTimestamp = performance.now()
          const endMemoryUsage = process.memoryUsage().heapUsed;
          executionTime = endTimestamp - startTimestamp;
          memoryUsage = endMemoryUsage - startMemoryUsage;
        `;
        vm.runInContext(script, sandbox);
        results.push({ results: sandbox.results, log: sandbox.log, executionTime: sandbox.executionTime, memoryUsage: sandbox.memoryUsage });
      }

      const deepEqual = (a, b) => {
        if (a === b) return true;
        if (typeof a !== typeof b) return false;
        if (typeof a !== "object" || a === null || b === null) return a === b;
        if (Array.isArray(a)) {
          if (!Array.isArray(b) || a.length !== b.length) return false;
          for (let i = 0; i < a.length; i++) {
            if (!deepEqual(a[i], b[i])) return false;
          }
          return true;
        }
        const keysA = Object.keys(a);
        const keysB = Object.keys(b);
        if (keysA.length !== keysB.length) return false;
        for (const key of keysA) {
          if (!keysB.includes(key) || !deepEqual(a[key], b[key])) return false;
        }
        return true;
      };

      const assessmentResult = {
        inputs: question.inputsOutputs.map((input) => input.inputs),
        outputs: results.map((result) => result.results),
        executionTimes: results.map((result) => result.executionTime),
        memoryUsages: results.map((result) => result.memoryUsage),
        logs: results.map((result) => result.log),
        expectedOutputs: question.inputsOutputs.map((input) => input.output),
        isCorrect: results.map((result, i) => deepEqual(result.results, question.inputsOutputs[i].output)),
      };

      res.status(200).json(assessmentResult);
    } catch (error) {
      console.error(error);
      if (error.message) {
        res.status(200).json({ error: error.message }); // Send error message to client
      } else {
        res.status(500).json({ error: "Error executing code." }); // Fallback error message
      }
    }
  }
};
